internal/jimmhttp/migrationproxy_handler.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2025 Canonical.
+
+package jimmhttp
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-chi/chi/v5"
+	jujuparams "github.com/juju/juju/rpc/params"
+
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/jimm"
+	"github.com/canonical/jimm/v3/internal/middleware"
+	"github.com/canonical/jimm/v3/internal/rpc"
+)
+
+// MigrationProxyHandler is an handler that provides proxying,
+// specifically for HTTP-based model migration endpoints. These
+// are used primarily to send data like charms/resources/tools
+// to the target controller during a model migration.
+// It differs slighly from the regular HTTPProxyHandler in that it
+// 1. Uses a custom HTTP header to pass the model UUID instead of extracting it
+// from the URL path.
+// 2. Performs different checks to ensure the model is in a
+// migration state before proxying the request to the controller.
+// 3. Requires the user to be a JIMM admin, rather than just a model writer.
+type MigrationHTTPProxyHandler struct {
+	Router *chi.Mux
+	jimm   *jimm.JIMM
+}
+
+const (
+	// all endpoints managed by this handler
+	migrationProxyEndpoints = "/*"
+)
+
+// NewMigrationHTTPProxyHandler creates a model migration proxy http handler.
+func NewMigrationHTTPProxyHandler(jimm *jimm.JIMM) *MigrationHTTPProxyHandler {
+	return &MigrationHTTPProxyHandler{Router: chi.NewRouter(), jimm: jimm}
+}
+
+// Routes returns the grouped routers routes with group specific middlewares.
+func (hph *MigrationHTTPProxyHandler) Routes() chi.Router {
+	hph.SetupMiddleware()
+	hph.Router.HandleFunc(migrationProxyEndpoints, hph.ProxyHTTP)
+	return hph.Router
+}
+
+// SetupMiddleware applies authn and authz middlewares.
+func (hph *MigrationHTTPProxyHandler) SetupMiddleware() {
+	hph.Router.Use(func(h http.Handler) http.Handler {
+		return middleware.AuthenticateWithSessionTokenViaBasicAuth(h, hph.jimm.LoginManager())
+	})
+	hph.Router.Use(middleware.AuthorizeUserAsJIMMAdmin)
+}
+
+// ProxyHTTP extracts the model uuid from an HTTP header to proxy the request to the right controller.
+func (hph *MigrationHTTPProxyHandler) ProxyHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
```

Comment:
Lil debug line saying this is starting the path (`req.URL.Path`) would be good. If we ever had issues in a prod deploy, we've got no visibility as to what endpoint has been hit / not been hit.